// @nestjs-repo/logger/dist/logger.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loggerConfig = loggerConfig;
const node_path_1 = require("node:path");
const rotating_file_stream_1 = require("rotating-file-stream");
const isProd = process.env.NODE_ENV === 'production';
const formatDate = (date) => {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}${month}${day}`;
};
const formatDateWithDash = (date) => {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
};
const formatDateTime = (date) => {
    const hours = String(date.getHours()).padStart(2, '0');
    const minutes = String(date.getMinutes()).padStart(2, '0');
    const seconds = String(date.getSeconds()).padStart(2, '0');
    return `${formatDateWithDash(date)} ${hours}:${minutes}:${seconds}`;
};
const serializers = {
    req(req) {
        const rawReq = req.raw;
        return {
            time: formatDateTime(new Date()),
            url: req.url,
            method: req.method,
            params: rawReq.params,
            query: rawReq.query,
            body: rawReq.body,
        };
    },
    res(res) {
        return {
            status: res.statusCode,
        };
    },
    err(err) {
        return err.type === 'Error' ? null : err;
    },
};
function loggerConfig() {
    if (isProd) {
        const stream = (0, rotating_file_stream_1.createStream)((time) => {
            const date = time
                ? time instanceof Date
                    ? time
                    : new Date(time)
                : new Date();
            return `${formatDate(date)}.log`;
        }, {
            interval: '1d',
            path: (0, node_path_1.join)(process.cwd(), 'logs'),
            maxFiles: 30,
            compress: false,
        });
        return {
            pinoHttp: [{ level: 'error', serializers }, stream],
        };
    }
    return {
        pinoHttp: {
            level: 'debug',
            serializers,
            transport: {
                target: 'pino-pretty',
                options: {
                    colorize: true,
                    translateTime: 'SYS:yyyy-mm-dd HH:MM:ss',
                    ignore: 'pid,hostname',
                },
            },
        },
    };
}
