// @nestjs-repo/logger/src/logger.config.ts
import { Request } from 'express';
import {
  SerializedError,
  SerializedRequest,
  SerializedResponse,
} from 'pino-std-serializers';
import { join } from 'node:path';
import { Params } from 'nestjs-pino';
import { createStream } from 'rotating-file-stream';

const isProd = process.env.NODE_ENV === 'production';

// 格式化日期为 YYYYMMDD（用于文件名）
const formatDate = (date: Date): string => {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}${month}${day}`;
};

// 格式化日期为 YYYY-MM-DD（用于日志内容）
const formatDateWithDash = (date: Date): string => {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
};

// 格式化时间为 YYYY-MM-DD HH:mm:ss
const formatDateTime = (date: Date): string => {
  const hours = String(date.getHours()).padStart(2, '0');
  const minutes = String(date.getMinutes()).padStart(2, '0');
  const seconds = String(date.getSeconds()).padStart(2, '0');
  return `${formatDateWithDash(date)} ${hours}:${minutes}:${seconds}`;
};

// 通用的序列化器配置
const serializers = {
  req(req: SerializedRequest) {
    const rawReq = req.raw as Request;
    return {
      time: formatDateTime(new Date()),
      url: req.url,
      method: req.method,
      params: rawReq.params,
      query: rawReq.query,
      body: rawReq.body as Record<string, unknown>,
    };
  },
  res(res: SerializedResponse) {
    return {
      status: res.statusCode,
    };
  },
  err(err: SerializedError) {
    return err.type === 'Error' ? null : err;
  },
};

export function loggerConfig(): Params {
  if (isProd) {
    // 生产环境：按日期轮转的文件日志
    const stream = createStream(
      (time) => {
        const date = time
          ? time instanceof Date
            ? time
            : new Date(time)
          : new Date();
        return `${formatDate(date)}.log`;
      },
      {
        interval: '1d',
        path: join(process.cwd(), 'logs'),
        maxFiles: 30,
        compress: false,
      },
    );

    return {
      pinoHttp: [{ level: 'error', serializers }, stream],
    };
  }

  // 开发环境：彩色控制台输出
  return {
    pinoHttp: {
      level: 'debug',
      serializers,
      transport: {
        target: 'pino-pretty',
        options: {
          colorize: true,
          translateTime: 'SYS:yyyy-mm-dd HH:MM:ss',
          ignore: 'pid,hostname',
        },
      },
    },
  };
}
