// enum-plus/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enum = exports.defaultLocalize = exports.ENUM_OPTIONS = exports.IS_ENUM = exports.IS_ENUM_ITEMS = exports.IS_ENUM_ITEM = exports.META = exports.NAMED = exports.LABELS = exports.VALUES = exports.ITEMS = exports.KEYS = exports.version = void 0;
var version_1 = require("./version");
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return version_1.version; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "KEYS", { enumerable: true, get: function () { return utils_1.KEYS; } });
Object.defineProperty(exports, "ITEMS", { enumerable: true, get: function () { return utils_1.ITEMS; } });
Object.defineProperty(exports, "VALUES", { enumerable: true, get: function () { return utils_1.VALUES; } });
Object.defineProperty(exports, "LABELS", { enumerable: true, get: function () { return utils_1.LABELS; } });
Object.defineProperty(exports, "NAMED", { enumerable: true, get: function () { return utils_1.NAMED; } });
Object.defineProperty(exports, "META", { enumerable: true, get: function () { return utils_1.META; } });
Object.defineProperty(exports, "IS_ENUM_ITEM", { enumerable: true, get: function () { return utils_1.IS_ENUM_ITEM; } });
Object.defineProperty(exports, "IS_ENUM_ITEMS", { enumerable: true, get: function () { return utils_1.IS_ENUM_ITEMS; } });
Object.defineProperty(exports, "IS_ENUM", { enumerable: true, get: function () { return utils_1.IS_ENUM; } });
Object.defineProperty(exports, "ENUM_OPTIONS", { enumerable: true, get: function () { return utils_1.ENUM_OPTIONS; } });
Object.defineProperty(exports, "defaultLocalize", { enumerable: true, get: function () { return utils_1.defaultLocalize; } });
var enum_1 = require("./enum");
Object.defineProperty(exports, "Enum", { enumerable: true, get: function () { return enum_1.Enum; } });
